In [1]:
# =============================================================================
# STEP 1: LOAD PROCESSED DATASET
# =============================================================================

import pandas as pd
import numpy as np


# Load cleaned dataset
df = pd.read_csv(
    "../../data/processed/breast_cancer_cleaned.csv"
)


print("=" * 70)
print("PROCESSED DATASET")
print("=" * 70)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")


display(df.head())

PROCESSED DATASET
Rows    : 569
Columns : 31


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


In [2]:
# =============================================================================
# STEP 2: FEATURE AND TARGET SEPARATION
# =============================================================================


target_column = "target"


X = df.drop(
    target_column,
    axis=1
)


y = df[target_column]



print("=" * 70)
print("FEATURE MATRIX (X)")
print("=" * 70)

print(f"Rows    : {X.shape[0]}")
print(f"Columns : {X.shape[1]}")


print("\n" + "=" * 70)
print("TARGET VARIABLE (y)")
print("=" * 70)

print(f"Rows : {y.shape[0]}")

display(y.head())

FEATURE MATRIX (X)
Rows    : 569
Columns : 30

TARGET VARIABLE (y)
Rows : 569


0    0
1    0
2    0
3    0
4    0
Name: target, dtype: int64

In [3]:
# =============================================================================
# STEP 3: VERIFY TARGET DISTRIBUTION
# =============================================================================


print("=" * 70)
print("TARGET DISTRIBUTION")
print("=" * 70)


display(
    y.value_counts()
)


print("\nClass Percentage")


display(
    (y.value_counts(normalize=True)*100)
    .round(2)
)

TARGET DISTRIBUTION


target
1    357
0    212
Name: count, dtype: int64


Class Percentage


target
1    62.74
0    37.26
Name: proportion, dtype: float64

In [4]:
# =============================================================================
# STEP 4: FEATURE TYPE CHECK
# =============================================================================


print("=" * 70)
print("FEATURE DATA TYPES")
print("=" * 70)


display(
    X.dtypes.to_frame(
        name="Data Type"
    )
)

FEATURE DATA TYPES


,Data Type
mean radius,float64
mean texture,float64
mean perimeter,float64
mean area,float64
mean smoothness,float64
mean compactness,float64
mean concavity,float64
mean concave points,float64
mean symmetry,float64
mean fractal dimension,float64


In [5]:
# =============================================================================
# STEP 3: TRAIN-TEST SPLIT
# =============================================================================


from sklearn.model_selection import train_test_split


print("=" * 70)
print("TRAIN-TEST SPLIT")
print("=" * 70)


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


print("Training Data")
print("-" * 70)

print(f"X_train shape : {X_train.shape}")
print(f"y_train shape : {y_train.shape}")


print("\nTesting Data")
print("-" * 70)

print(f"X_test shape  : {X_test.shape}")
print(f"y_test shape  : {y_test.shape}")

TRAIN-TEST SPLIT
Training Data
----------------------------------------------------------------------
X_train shape : (455, 30)
y_train shape : (455,)

Testing Data
----------------------------------------------------------------------
X_test shape  : (114, 30)
y_test shape  : (114,)


In [6]:
# =============================================================================
# CLASS DISTRIBUTION AFTER SPLIT
# =============================================================================


print("=" * 70)
print("TRAINING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_train.value_counts(normalize=True)*100)
    .round(2)
)



print("\n" + "=" * 70)
print("TESTING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_test.value_counts(normalize=True)*100)
    .round(2)
)

TRAINING TARGET DISTRIBUTION


target
1    62.64
0    37.36
Name: proportion, dtype: float64


TESTING TARGET DISTRIBUTION


target
1    63.16
0    36.84
Name: proportion, dtype: float64

In [7]:
# =============================================================================
# STEP 4: FEATURE SCALING
# =============================================================================


from sklearn.preprocessing import StandardScaler


print("=" * 70)
print("FEATURE SCALING USING STANDARD SCALER")
print("=" * 70)


# Initialize scaler
scaler = StandardScaler()


# Fit only on training data
X_train_scaled = scaler.fit_transform(X_train)


# Transform test data using same scaler
X_test_scaled = scaler.transform(X_test)



print("Scaling completed successfully")


print("\n" + "=" * 70)
print("SHAPE AFTER SCALING")
print("=" * 70)


print(f"X_train_scaled shape : {X_train_scaled.shape}")
print(f"X_test_scaled shape  : {X_test_scaled.shape}")

FEATURE SCALING USING STANDARD SCALER
Scaling completed successfully

SHAPE AFTER SCALING
X_train_scaled shape : (455, 30)
X_test_scaled shape  : (114, 30)


In [8]:
# =============================================================================
# CONVERT SCALED DATA BACK TO DATAFRAME
# =============================================================================


X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns
)


X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns
)



print("=" * 70)
print("SCALED TRAIN DATA PREVIEW")
print("=" * 70)


display(X_train_scaled.head())

SCALED TRAIN DATA PREVIEW


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,-1.072001,-0.658425,-1.088080,-0.939274,-0.135940,-1.008718,-0.968359,-1.102032,0.281062,-0.113231,...,-1.034094,-0.623497,-1.070773,-0.876534,-0.169982,-1.038836,-1.078995,-1.350527,-0.352658,-0.541380
1,1.748743,0.066502,1.751157,1.745559,1.274468,0.842288,1.519852,1.994664,-0.293045,-0.320180,...,1.228342,-0.092833,1.187467,1.104386,1.517001,0.249655,1.178594,1.549916,0.191078,-0.173739
2,-0.974734,-0.931124,-0.997709,-0.867589,-0.613515,-1.138154,-1.092292,-1.243358,0.434395,-0.429247,...,-0.973231,-1.036772,-1.008044,-0.834168,-1.097823,-1.167260,-1.282241,-1.707442,-0.307734,-1.213033
3,-0.145103,-1.215186,-0.123013,-0.253192,0.664482,0.286762,-0.129729,-0.098605,0.555635,0.029395,...,-0.251266,-1.369643,-0.166633,-0.330292,0.234006,0.096874,-0.087521,-0.344838,0.242198,-0.118266
4,-0.771617,-0.081211,-0.803700,-0.732927,-0.672282,-1.006099,-0.798502,-0.684484,0.737495,-0.457213,...,-0.801135,0.079230,-0.824381,-0.741830,-0.911367,-0.984612,-0.933190,-0.777604,0.555118,-0.761639


In [9]:
# =============================================================================
# SCALING VERIFICATION
# =============================================================================


print("=" * 70)
print("MEAN AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.mean().round(2)
)



print("\n" + "=" * 70)
print("STANDARD DEVIATION AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.std().round(2)
)

MEAN AFTER SCALING


mean radius               -0.0
mean texture               0.0
mean perimeter            -0.0
mean area                 -0.0
mean smoothness            0.0
mean compactness          -0.0
mean concavity            -0.0
mean concave points        0.0
mean symmetry             -0.0
mean fractal dimension     0.0
radius error              -0.0
texture error             -0.0
perimeter error            0.0
area error                 0.0
smoothness error          -0.0
compactness error         -0.0
concavity error            0.0
concave points error      -0.0
symmetry error             0.0
fractal dimension error    0.0
worst radius               0.0
worst texture             -0.0
worst perimeter           -0.0
worst area                 0.0
worst smoothness           0.0
worst compactness         -0.0
worst concavity            0.0
worst concave points      -0.0
worst symmetry            -0.0
worst fractal dimension    0.0
dtype: float64


STANDARD DEVIATION AFTER SCALING


mean radius                1.0
mean texture               1.0
mean perimeter             1.0
mean area                  1.0
mean smoothness            1.0
mean compactness           1.0
mean concavity             1.0
mean concave points        1.0
mean symmetry              1.0
mean fractal dimension     1.0
radius error               1.0
texture error              1.0
perimeter error            1.0
area error                 1.0
smoothness error           1.0
compactness error          1.0
concavity error            1.0
concave points error       1.0
symmetry error             1.0
fractal dimension error    1.0
worst radius               1.0
worst texture              1.0
worst perimeter            1.0
worst area                 1.0
worst smoothness           1.0
worst compactness          1.0
worst concavity            1.0
worst concave points       1.0
worst symmetry             1.0
worst fractal dimension    1.0
dtype: float64

# ANOVA

## 1. Baseline

In [10]:
# =============================================================================
# ANOVA BASELINE FEATURE SELECTION (TOP-K = 5,10,15,20)
# =============================================================================

from sklearn.feature_selection import SelectKBest, f_classif
import pandas as pd


# Fit ANOVA once with all features
anova_selector = SelectKBest(
    score_func=f_classif,

    k="all"
)

anova_selector.fit(
    X_train_scaled,
    y_train
)


# Create score dataframe
anova_scores = pd.DataFrame({

    "Feature": X_train_scaled.columns,

    "ANOVA Score": anova_selector.scores_

})


# Sort by ANOVA importance
anova_scores = anova_scores.sort_values(
    by="ANOVA Score",
    ascending=False
).reset_index(drop=True)



# Top-k values
k_values = [5, 10, 15, 20]

anova_results = {}


for k in k_values:

    top_features = (
        anova_scores
        .head(k)["Feature"]
        .tolist()
    )


    # Boolean mask for selecting columns
    mask = X_train_scaled.columns.isin(
        top_features
    )


    X_train_selected = X_train_scaled.loc[
        :, mask
    ]

    X_test_selected = X_test_scaled.loc[
        :, mask
    ]


    anova_results[k] = {

        "features": top_features,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "scores": anova_scores.head(k)

    }


    print("="*70)
    print(f"TOP {k} ANOVA FEATURES")
    print("="*70)

    print(top_features)

    print("\nScores:")
    print(anova_scores.head(k))

TOP 5 ANOVA FEATURES
['worst concave points', 'worst perimeter', 'mean concave points', 'worst radius', 'mean perimeter']

Scores:
                Feature  ANOVA Score
0  worst concave points   795.223341
1       worst perimeter   719.598823
2   mean concave points   703.364031
3          worst radius   687.430367
4        mean perimeter   565.485755
TOP 10 ANOVA FEATURES
['worst concave points', 'worst perimeter', 'mean concave points', 'worst radius', 'mean perimeter', 'worst area', 'mean radius', 'mean area', 'mean concavity', 'worst concavity']

Scores:
                Feature  ANOVA Score
0  worst concave points   795.223341
1       worst perimeter   719.598823
2   mean concave points   703.364031
3          worst radius   687.430367
4        mean perimeter   565.485755
5            worst area   541.713694
6           mean radius   523.613907
7             mean area   469.516453
8        mean concavity   416.374874
9       worst concavity   365.747389
TOP 15 ANOVA FEATURES
['worst

In [11]:
# =============================================================================
# BASELINE MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {
    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        random_state=42,

        eval_metric="logloss",
        scale_pos_weight=(y_train == 0).sum() / (y_train == 1).sum()
    )
}

In [12]:
# =============================================================================
# MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

results = []

for k in k_values:

    Xtr = anova_results[k]["X_train"]
    Xts = anova_results[k]["X_test"]

    features = ", ".join(anova_results[k]["features"])

    for model_name, model in models.items():

        model.fit(Xtr, y_train)

        y_pred = model.predict(Xts)

        y_prob = model.predict_proba(Xts)[:,1]

        results.append({

            "Method": "ANOVA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(y_test, y_pred),

            "Precision": precision_score(y_test, y_pred),

            "Recall": recall_score(y_test, y_pred),

            "F1 Score": f1_score(y_test, y_pred),

            "ROC-AUC": roc_auc_score(y_test, y_prob)

        })

In [13]:
results_df = pd.DataFrame(results)

results_df

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,ANOVA,5,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.929825,0.984848,0.902778,0.942029,0.991071
1,ANOVA,5,Random Forest,"worst concave points, worst perimeter, mean co...",0.929825,0.957143,0.930556,0.943662,0.988261
2,ANOVA,5,XGBoost,"worst concave points, worst perimeter, mean co...",0.938596,0.957746,0.944444,0.951049,0.984458
3,ANOVA,10,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.912281,0.969697,0.888889,0.927536,0.991402
4,ANOVA,10,Random Forest,"worst concave points, worst perimeter, mean co...",0.956140,0.958904,0.972222,0.965517,0.990079
5,ANOVA,10,XGBoost,"worst concave points, worst perimeter, mean co...",0.973684,0.972603,0.986111,0.979310,0.992394
6,ANOVA,15,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.956140,0.985507,0.944444,0.964539,0.991402
7,ANOVA,15,Random Forest,"worst concave points, worst perimeter, mean co...",0.956140,0.971831,0.958333,0.965035,0.991733
8,ANOVA,15,XGBoost,"worst concave points, worst perimeter, mean co...",0.964912,0.959459,0.986111,0.972603,0.992394
9,ANOVA,20,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.964912,0.985714,0.958333,0.971831,0.995040


In [14]:
import os
os.makedirs("../../../results/breast_cancer/baseline", exist_ok=True)

results_df.to_csv(
    "../../../results/breast_cancer/baseline/anova_baseline_results.csv",
    index=False
)

## 2. DODA

In [15]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "--upgrade",
    "--force-reinstall",
    "--no-cache-dir",
    "git+https://github.com/anandha-3679/DODA.git"
])

0

In [16]:
import sys
import doda
from doda import DODASelector
import inspect

print("Python:", sys.executable)
print("DODA:", doda.__file__)
print("Selector:", inspect.getfile(DODASelector))

Python: c:\Users\johnm\msc_research\.venv\Scripts\python.exe
DODA: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\__init__.py
Selector: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\selector.py


In [17]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from doda import DODASelector

from doda.adapters import SklearnAdapter

from doda.knowledge.providers import JSONProvider

from doda.fusion import HadamardFusion

In [ ]:
# =============================================================================
# ANOVA + DODA FEATURE SELECTION
# =============================================================================

import pandas as pd
from sklearn.feature_selection import SelectKBest, f_classif

# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 10, 15, 20]


# -------------------------------------------------------------------------
# Clinical knowledge provider
# -------------------------------------------------------------------------

provider = JSONProvider(
    "../../../config/clinical_weights/breast_cancer.json"
)


# -------------------------------------------------------------------------
# Fusion method
# -------------------------------------------------------------------------

fusion = HadamardFusion()


# -------------------------------------------------------------------------
# Store DODA results
# -------------------------------------------------------------------------

doda_results = {}


# =============================================================================
# RUN DODA FOR EACH TOP-K
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"ANOVA + DODA \u2014 TOP {k} FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # ANOVA operator
    # ---------------------------------------------------------------------

    anova = SklearnAdapter(
        SelectKBest(
            score_func=f_classif,
            k=k
        )
    )


    # ---------------------------------------------------------------------
    # DODA selector
    # ---------------------------------------------------------------------

    selector = DODASelector(
        operators=[anova],
        provider=provider,
        fusion=fusion,
        top_k=k
    )


    # ---------------------------------------------------------------------
    # Fit DODA on TRAINING data
    # ---------------------------------------------------------------------

    X_train_selected = selector.fit_transform(
        X_train_scaled,
        y_train
    )


    # ---------------------------------------------------------------------
    # Transform TEST data
    # ---------------------------------------------------------------------

    X_test_selected = selector.transform(
        X_test_scaled
    )


    # ---------------------------------------------------------------------
    # Selected features
    # ---------------------------------------------------------------------

    selected_features = selector.get_selected_features()


    print("\nSelected Features:")
    print(selected_features)


    # ---------------------------------------------------------------------
    # Store everything needed for later evaluation
    # ---------------------------------------------------------------------

    doda_results[k] = {

        "selector": selector,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "features": selected_features,

        "raw_math_scores": selector.raw_math_scores_,

        "math_scores": selector.math_scores_,

        "clinical_weights": selector.clinical_weights_,

        "final_scores": selector.final_scores_
    }


print("\n" + "=" * 70)
print("DODA FEATURE SELECTION COMPLETED")
print("=" * 70)

JSONDecodeError: Illegal trailing comma before end of object: line 33 column 33 (char 787)

In [19]:
# CELL 2
# =============================================================================
# INSPECT MATHEMATICAL SCORES
# =============================================================================

k = 10

selector = doda_results[k]["selector"]

# -----------------------------------------------------------------------------
# Raw mathematical scores
# -----------------------------------------------------------------------------

raw_math_scores = pd.DataFrame(
    selector.raw_math_scores_.items(),
    columns=[
        "Feature",
        "Raw Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Normalized mathematical scores
# -----------------------------------------------------------------------------

normalized_math_scores = pd.DataFrame(
    selector.math_scores_.items(),
    columns=[
        "Feature",
        "Normalized Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Combine both
# -----------------------------------------------------------------------------

math_scores = raw_math_scores.merge(
    normalized_math_scores,
    on="Feature"
)

math_scores = math_scores.sort_values(
    by="Normalized Math Score",
    ascending=False
)

print("=" * 70)
print("MATHEMATICAL SCORES")
print("=" * 70)

display(math_scores)

NameError: name 'doda_results' is not defined

In [ ]:
# CELL 3
# =============================================================================
# INSPECT CLINICAL WEIGHTS
# =============================================================================

clinical_weights = pd.DataFrame(
    selector.clinical_weights_.items(),
    columns=[
        "Feature",
        "Clinical Weight"
    ]
)

clinical_weights = clinical_weights.sort_values(
    by="Clinical Weight",
    ascending=False
)

print("=" * 70)
print("CLINICAL WEIGHTS")
print("=" * 70)

display(clinical_weights)

In [ ]:
# CELL 4
# =============================================================================
# INSPECT FINAL DODA SCORES
# =============================================================================

final_scores = pd.DataFrame(
    selector.final_scores_.items(),
    columns=[
        "Feature",
        "Final DODA Score"
    ]
)

final_scores = final_scores.sort_values(
    by="Final DODA Score",
    ascending=False
)

print("=" * 70)
print("FINAL DODA SCORES")
print("=" * 70)

display(final_scores)

In [ ]:
selector.get_selected_features()

In [ ]:
comparison = selector.compare_scores()

comparison

In [ ]:
# =============================================================================
# DODA MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


doda_model_results = []


for k in k_values:

    Xtr = doda_results[k]["X_train"]

    Xts = doda_results[k]["X_test"]

    features = ", ".join(
        doda_results[k]["features"]
    )


    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        doda_model_results.append({

            "Method": "ANOVA + DODA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


doda_results_df = pd.DataFrame(
    doda_model_results
)


display(doda_results_df)

In [ ]:
# =============================================================================
# COMBINED COMPARISON
# =============================================================================

comparison_df = pd.concat(
    [
        results_df,
        doda_results_df
    ],
    ignore_index=True
)


display(
    comparison_df
)

In [ ]:
for k in k_values:

    selector = doda_results[k]["selector"]

    raw_rank = [
        x[0]
        for x in sorted(
            selector.raw_math_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    doda_rank = [
        x[0]
        for x in sorted(
            selector.final_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    print("\n" + "=" * 70)
    print(f"TOP {k}")
    print("=" * 70)

    print("Raw Math Ranking:")
    print(raw_rank)

    print("\nDODA Ranking:")
    print(doda_rank)

    print("\nTop-K Same:",
          raw_rank[:k] == doda_rank[:k])

# 3. ANOVA + DODA: Model Evaluation & Interpretation

## Overall Predictive Performance

The executed notebook contains complete ANOVA baseline model-evaluation results for Top-K = 5, 10, 15, and 20 using Logistic Regression, Random Forest, and XGBoost. However, the ANOVA + DODA feature-selection cell did not complete successfully. Its executed output reports `JSONDecodeError: Illegal trailing comma before end of object: line 33 column 33 (char 787)`. The following DODA inspection cell then reports `NameError: name 'doda_results' is not defined`, and the DODA model-evaluation and combined-comparison cells have no executed outputs. Therefore, a direct metric-by-metric comparison between the ANOVA baseline and DODA cannot yet be made from this notebook.

For the ANOVA baseline, the observed model-evaluation results were as follows:

| Top-K | Model | Accuracy | Precision | Recall | F1-score | ROC-AUC |
|---:|---|---:|---:|---:|---:|---:|
| 5 | Logistic Regression | 0.929825 | 0.984848 | 0.902778 | 0.942029 | 0.991071 |
| 5 | Random Forest | 0.929825 | 0.957143 | 0.930556 | 0.943662 | 0.988261 |
| 5 | XGBoost | 0.938596 | 0.957746 | 0.944444 | 0.951049 | 0.984458 |
| 10 | Logistic Regression | 0.912281 | 0.969697 | 0.888889 | 0.927536 | 0.991402 |
| 10 | Random Forest | 0.956140 | 0.958904 | 0.972222 | 0.965517 | 0.990079 |
| 10 | XGBoost | 0.973684 | 0.972603 | 0.986111 | 0.979310 | 0.992394 |
| 15 | Logistic Regression | 0.956140 | 0.985507 | 0.944444 | 0.964539 | 0.991402 |
| 15 | Random Forest | 0.956140 | 0.971831 | 0.958333 | 0.965035 | 0.991733 |
| 15 | XGBoost | 0.964912 | 0.959459 | 0.986111 | 0.972603 | 0.992394 |
| 20 | Logistic Regression | 0.964912 | 0.985714 | 0.958333 | 0.971831 | 0.995040 |
| 20 | Random Forest | 0.947368 | 0.971429 | 0.944444 | 0.957746 | 0.993717 |
| 20 | XGBoost | 0.938596 | 0.957746 | 0.944444 | 0.951049 | 0.995370 |

Within the ANOVA baseline alone, XGBoost at Top-K = 10 produced the highest observed accuracy, recall, and F1-score, while the highest observed ROC-AUC was obtained by XGBoost at Top-K = 20. Logistic Regression at Top-K = 5 and Top-K = 20 produced the highest observed precision. Because no executed DODA performance table is present, the notebook does not show whether DODA improved, maintained, or reduced accuracy, precision, recall, F1-score, or ROC-AUC relative to ANOVA.

## Effect of DODA on Feature Selection

The executed ANOVA baseline feature-selection output ranks the Top-20 features as: `worst concave points`, `worst perimeter`, `mean concave points`, `worst radius`, `mean perimeter`, `worst area`, `mean radius`, `mean area`, `mean concavity`, `worst concavity`, `mean compactness`, `worst compactness`, `area error`, `radius error`, `perimeter error`, `worst texture`, `worst symmetry`, `worst smoothness`, `concave points error`, and `mean texture`. The Top-K subsets are nested within this ordering for K = 5, 10, 15, and 20.

The DODA-selected features are not present in executed output because the DODA selection cell failed before creating `doda_results`. Consequently, this notebook does not provide evidence for which ANOVA-selected features were retained by DODA, which features changed rank after DODA, or whether any features were promoted or demoted by the DODA fusion step.

The available clinical-weight configuration gives the highest weights to `worst radius`, `worst perimeter`, `worst area`, `worst concavity`, and `worst concave points` (1.0), and high weights to `mean radius`, `mean perimeter`, `mean area`, `mean concavity`, and `mean concave points` (0.9). These clinically weighted features are already strongly represented in the ANOVA Top-10 output. However, because the executed DODA ranking is missing, it cannot yet be concluded whether DODA meaningfully changed the selected feature space or altered the ranking in favour of clinically weighted features.

## Current Conclusion

The observed results show that the breast cancer dataset contains 569 rows, 30 feature columns, and one target column, with a target distribution of 357 cases in class 1 and 212 cases in class 0. After the train-test split, the training set contains 455 rows and the test set contains 114 rows. The ANOVA baseline produces high predictive performance across the evaluated models and Top-K settings, with all observed ROC-AUC values ranging from 0.984458 to 0.995370.

The current interpretation is that ANOVA-selected features provide a strong baseline for this breast cancer classification experiment. The strongest ANOVA-ranked features are dominated by worst and mean measures of concave points, perimeter, radius, area, and concavity, which correspond to features assigned high clinical weights in the repository configuration.

What cannot yet be concluded is whether DODA improves, maintains, or reduces predictive performance compared with ANOVA, because no executed DODA model-evaluation results are available in the notebook. It also cannot yet be concluded that DODA produces a more clinically aligned feature subset, changes feature ranks meaningfully, or is superior to the ANOVA baseline. Based only on the actual executed outputs, the notebook currently supports conclusions about the ANOVA baseline, but not about the effectiveness of ANOVA + DODA for this breast cancer experiment.


# 3. ANOVA + DODA: Model Evaluation & Interpretation

## Overall Predictive Performance

The ANOVA baseline produced strong performance across all Top-K settings and all three classifiers. ROC-AUC was consistently high, ranging from 0.984458 to 0.995370, indicating that the statistically selected features carried substantial discriminatory information for the breast cancer classification task. The strongest single ANOVA result by accuracy and F1-score was XGBoost at Top-K = 10, with accuracy = 0.973684, recall = 0.986111, F1-score = 0.979310, and ROC-AUC = 0.992394. The highest ANOVA ROC-AUC was obtained by XGBoost at Top-K = 20 with ROC-AUC = 0.995370.

DODA changed performance in a model- and Top-K-dependent way rather than producing a uniform increase or decrease. At Top-K = 5, DODA improved Logistic Regression accuracy from 0.929825 to 0.938596, recall from 0.902778 to 0.916667, and F1-score from 0.942029 to 0.949640, while ROC-AUC changed only slightly from 0.991071 to 0.990741. Random Forest also improved at Top-K = 5 in accuracy, precision, and F1-score, although ROC-AUC decreased from 0.988261 to 0.982804. In contrast, XGBoost at Top-K = 5 was reduced by DODA across accuracy, precision, recall, F1-score, and ROC-AUC.

At Top-K = 10, DODA maintained exactly the same Logistic Regression performance as ANOVA, but reduced Random Forest and XGBoost performance slightly. For Random Forest, accuracy decreased from 0.956140 to 0.947368 and F1-score from 0.965517 to 0.958333. For XGBoost, accuracy decreased from 0.973684 to 0.964912 and F1-score from 0.979310 to 0.972603, while recall remained unchanged at 0.986111.

At Top-K = 15, DODA maintained the Logistic Regression result exactly and maintained the XGBoost classification metrics, with XGBoost ROC-AUC increasing slightly from 0.992394 to 0.992725. The clearest DODA improvement at this feature size was for Random Forest: accuracy increased from 0.956140 to 0.964912, recall from 0.958333 to 0.972222, and F1-score from 0.965035 to 0.972222, although ROC-AUC decreased from 0.991733 to 0.989087.

At Top-K = 20, DODA again maintained Logistic Regression exactly. Random Forest remained comparable overall, with the same accuracy, improved recall and ROC-AUC, but lower precision. XGBoost improved at Top-K = 20, with accuracy increasing from 0.938596 to 0.947368, recall from 0.944444 to 0.958333, F1-score from 0.951049 to 0.958333, and ROC-AUC from 0.995370 to 0.996032.

Overall, DODA improved performance most clearly for Logistic Regression and Random Forest at Top-K = 5, Random Forest at Top-K = 15, and XGBoost at Top-K = 20. It maintained comparable performance for Logistic Regression at Top-K = 10, 15, and 20, and for XGBoost at Top-K = 15. It reduced performance for XGBoost at Top-K = 5 and 10, and for Random Forest at Top-K = 10. These results indicate that DODA did not uniformly outperform ANOVA, but it often preserved comparable predictive performance while altering the feature ranking.

## Effect of DODA on Feature Selection

The ANOVA ranking was led by `worst concave points`, `worst perimeter`, `mean concave points`, `worst radius`, `mean perimeter`, `worst area`, `mean radius`, `mean area`, `mean concavity`, and `worst concavity`. This shows that the baseline statistical selection was already dominated by lesion size, boundary, and concavity-related measurements.

The DODA ranking retained the same broad feature family but reordered several features after applying the configured clinical weights. The DODA Top-10 ranking was `worst concave points`, `worst perimeter`, `worst radius`, `mean concave points`, `worst area`, `mean perimeter`, `mean radius`, `mean area`, `mean concavity`, and `worst concavity`. Thus, the Top-10 feature set remained the same as ANOVA, but the internal ordering changed.

Several high-ranking features were retained by both approaches across the evaluated Top-K values, including `worst concave points`, `worst perimeter`, `mean concave points`, `worst radius`, `mean perimeter`, `worst area`, `mean radius`, `mean area`, `mean concavity`, and `worst concavity`. These retained features also have high clinical weights in the breast cancer knowledge file: `worst concave points`, `worst perimeter`, `worst radius`, `worst area`, and `worst concavity` each have weight 1.0, while `mean concave points`, `mean perimeter`, `mean radius`, `mean area`, and `mean concavity` each have weight 0.9.

DODA promoted clinically higher-weighted features within the ranking. `worst radius` moved from ANOVA rank 4 to DODA rank 3, and `worst area` moved from ANOVA rank 6 to DODA rank 5, entering the Top-5 DODA subset. This replaced `mean perimeter`, which was rank 5 under ANOVA but rank 6 under DODA. At Top-K = 15, `worst compactness` moved above `mean compactness`, consistent with their configured clinical weights of 0.8 and 0.7 respectively. At Top-K = 20, `mean texture` moved above `concave points error`; their clinical weights are 0.5 and 0.4 respectively.

DODA also demoted some features relative to their ANOVA positions. `mean concave points` moved from rank 3 to rank 4, `mean perimeter` from rank 5 to rank 6, `mean compactness` from rank 11 to rank 12, and `concave points error` from rank 19 to rank 20. These changes do not remove the clinically relevant core of the ANOVA ranking, but they show how DODA shifts preference toward features with stronger configured clinical support when statistical scores are close enough for the weighting to affect order.

The main feature-selection effect in this notebook is therefore refinement rather than wholesale replacement. DODA did not introduce a completely different feature family; instead, it recovered or moved upward clinically weighted features such as `worst area`, `worst radius`, and `worst compactness`. This aligns with the purpose of DODA: to combine statistical evidence with clinical knowledge so that clinically important features are not ranked lower solely because of the mathematical selection score.

## Current Conclusion

The ANOVA baseline demonstrates that a compact subset of statistically selected breast cancer features can achieve strong predictive performance on this dataset. The strongest ANOVA features are concentrated around worst and mean measurements of concave points, perimeter, radius, area, and concavity, all of which are clinically weighted highly in the configured knowledge file.

DODA changes the selected feature space mainly by reordering features within this same clinically meaningful group. Its most visible effect is to promote features with higher clinical weights, including `worst radius`, `worst area`, and `worst compactness`, while demoting lower-weighted or comparatively less clinically prioritised alternatives in some Top-K settings.

Predictive performance under DODA is broadly comparable to the ANOVA baseline, but not uniformly superior. DODA improves some model and Top-K combinations, maintains others, and reduces others. This is important because the results suggest that clinical weighting can refine the selected feature ranking without causing a general collapse in predictive performance, but they do not show consistent performance gains across all classifiers.

For this breast cancer experiment, the results support the idea that clinical knowledge can refine statistically selected features by altering the rank order toward clinically important variables while often maintaining similar predictive performance. However, this single dataset cannot establish that DODA generalizes across diseases, datasets, or cohorts. It also cannot establish statistical significance, because no significance testing was performed in the notebook. The current conclusion is therefore limited to this breast cancer experiment: DODA provides a clinically informed refinement of the ANOVA-selected feature space, with mostly comparable and occasionally improved predictive performance, but further datasets and formal validation are needed before making broader claims.

